doc/designs/disable-stale-users.md

```diff
@@ -0,0 +1,291 @@
+changed regularly. A user account whose associated password is past its due
+change date can be considered stale and therefore be disabled by the system.
+In this case, a grace period must be provided between password expiration and
+the actual account locking event.
+Total writes to LDAP using this approach due to the disable stale users tooling
+are minimal (only nsAccountLock). However, password-change related writes are
+needed every 90 day (by default).
+Each password change updates the krbLastPwdChange, krbPasswordExpiration,
+userPassword, krbPrincipalKey, and krbExtraData attributes.
+
+The Disable Stale Users (DSU) tooling proposes to make sure no non-locked
+account has a password expired since N days, N being 7 by default.
+
+### Environments with no mandatory password changes
+
+For environments where user passwords never expire, a new mechanism must be
+added to FreeIPA plugins ipa-kdb and ipa-lockout to account for each user
+account authentication events. Since updating a LDAP user account attribute on
+each authentication event must be avoided as the update frequency could easily
+be high, (see https://pagure.io/freeipa/issue/5313 for background
+information), a coarse update mechanism is proposed.
+
+An attribute containing a timestamp is added to each user:
+coarseSuccessfulAuthTime. The timestamp contained therein is not precise.
+It will be added at account creation time (set to current date).
+The timestamp is updated by IPA plugins whenever a successful
+authentication event happens, but only if the previous timestamp is
+sufficiently in the past.
+
+Three ipaConfigString attributes must be added to the cn=ipaConfig schema:
+* the first one to activate the mechanism (default: False):
+  DSU:RecordSuccessfulAuth
+* the second one contains the fixed time period (default: 28 days):
+  DSU:RecordSucessfulAuth-TimeInterval
+* the third one contains a random time period to be added (default: 7 days):
+  DSU:RecordSucessfulAuth-RandomizedTimeInterval
+
+The time period to take into account is 28+7 so 35 days.
+
+As a rule of thumb, if the requirement is to disable users not seen for 90 days
+maximum, updating the attribute every 35 days is good enough.
+At worst, the user would have authenticated the first day and at least once
+during this 35 day period, say day #29. The attribute would therefore contain
+the date from the first day, as day #29 would have been considered too close to
+the existing timestamp to warrant updating it. The user account would then be
+disabled after 61 days of inactivity which fits (loosely) the requirement.
+In short, having a 90-day inactive user disabling requirement would by default
+disable accounts inactive for 55 (90-35) to 90 days. 
+
+Updating this (replicated) timestamp attribute every ~30 days is in fact better
+than mandating password changes every 90 days as the latter would imply
+updating five (replicated) attributes at least once in that 90-day period.
```

Comment:
I'm not sure I'd characterize it as better because password changes may be due to compliance so it out of the hands of the user.

Reply:
Fixed, thanks.